### DISCUSSION TITLE
Tooltip not convinient

### DISCUSSION THREAD

**Discussion by vallsv** [silx-kit/h5web#1110] [Q&A]
The actual tooltip is not convenient for my use case.

- It is constrained inside the plot, you can see the tooltip is cut
- it moves from right to left to optimize the location in the plot ; but i really don't want blinking of the tooltip content, the screen is not limited to a widget.

![image](https://user-images.githubusercontent.com/7579321/167600115-c5488be0-9785-46b3-a003-6b1f2bd32e70.png)
![image](https://user-images.githubusercontent.com/7579321/167600696-e7ff0bd2-377f-46ab-b724-a916bfa6f90b.png)

Ideally i would prefer to have mouse event (enter, move, leave), and a way to handle the tooltip container so what i can reuse the tooltip from the react bootstrap framework.

**Comment by axelboc** [comment]
I've activated the discussions and converted this issue into a discussion as it sounds like you're looking for a way to implement your own tooltip.

You'll need to render a ThreeJS `mesh` inside the visualization that takes up the entire canvas, like we do in https://github.com/silx-kit/h5web/blob/main/packages/lib/src/vis/shared/TooltipMesh.tsx. Then, you can add event listeners to the mesh and use [@visx/tooltip](https://airbnb.io/visx/docs/tooltip) (or your own React states) to manage the state of the tooltip (visible/hidden, position, data).

Once you have the state of the tooltip, it is up to you to render the actual tooltip. You'll just have to render it inside an `Html` element (since the tooltip is made of HTML elements and not R3F objects). In `TooltipOverlay`, we use `Overlay`, which uses `Html` internally, but this is because we _want_ the tooltip to be bound to the canvas, so we render it inside a container that takes up the entire canvas. If you don't need that, then `Html` is enough.

Now the slightly tricky bit is the overflow: `Html` renders its children inside a `div` and, by default, appends this `div` to the R3F canvas' wrapper `div` element, `r3fRoot`. However, this element hides its overflow, so if you don't want your tooltip to be cropped, you'll have to tell `Html` to render in `r3fRoot`'s parent `div`: `canvasWrapper`. Fortunately, this is exactly what we do in `AxisSystem`: https://github.com/silx-kit/h5web/blob/13a6f00134f41d00498f5a6417d9b7e22e9b2dbe/packages/lib/src/vis/shared/AxisSystem.tsx#L31.

**Reply by vallsv** [reply]
Could it be possible to create an example? Because i don't think i understand.

You talk about `TooltipMesh` events? Is the mouse coords will be properly exposed as data coords?

On side note, if the way to put tooltip at a location in the plot is convenient, i also think about using it to put marker pins on top of the plot (which is not annotations -> which are inside the plot). In this case i have to know how to check a position is actually part of the plot. This could be another discussion.

**Reply by vallsv** [reply]
BTW it's not urgent, i can find my way with something good enough for now.

**Reply by axelboc** [reply]
> Ideally i would prefer to have mouse event (enter, move, leave), and a way to handle the tooltip container so what i can reuse the tooltip from the react bootstrap framework.

If you want the above, then `TooltipMesh` is what you need to re-implement in daiquiri. In other words, it _is_ the example ^^

We register R3F events on the mesh here: https://github.com/silx-kit/h5web/blob/main/packages/lib/src/vis/shared/TooltipMesh.tsx#L82.

We convert the world coordinates to data coordinates here: https://github.com/silx-kit/h5web/blob/main/packages/lib/src/vis/shared/TooltipMesh.tsx#L44

**Reply by vallsv** [reply]
I am not sure i understand what you mean.

I am already using `TooltipMesh`. You propose me to copy paste this module into Daiquiri? Is everything inside is public?

To me it's an ok solution for a very short term, but such code have to be provided by h5web.
For example the h5web `Selection` API sounds like more convenient to use.

Furthermore could it be possible to provide mouse callbacks anyway?
That's something we will need to have at some point anyway. For example here is a crosshair propagated to side plots.

![image](https://user-images.githubusercontent.com/7579321/167805615-78df1be3-fc26-4519-aa15-c252903b1e2c.png)

**Reply by axelboc** [reply]
At its core, the H5Web library sets up a ThreeJS canvas with React Three Fiber and provides an axis system with some scales to go from the ThreeJS world space to the data space and back. React Three Fiber and our axis system are already powerful abstractions that hugely simplify dealing with ThreeJS and WebGL!

The components we've built so far in the H5Web library, like `TooltipMesh`, respond to the needs of the H5Web viewer, `@h5web/app`, so they are not necessarily the right abstractions for you to use in daiquiri, as you've pointed out. Perhaps some of our components will evolve to cover daiquiri's needs eventually, but **the most productive way to get there is to first prototype something in daiquiri**, and our job is to help you with doing just that. This is what Thomas has been doing, for instance, with the ROIs or synchronisation of the camera between two plots. Once you have something that roughly covers your needs, we can discuss bringing some parts of it into H5Web so it can be maintained and improved there and benefit future projects, and so your code in daiquiri can be simplified.

Obviously our axis system abstraction is sometimes not sufficient. This is the case when it comes to drawing SVG elements in data space on top of the canvas, for instance, but we're currently working on a solution to this problem. When it comes to the tooltip, however, React Three Fiber, our axis system utilities, and our low-level `Html` component should already allow you to prototype what you need, as I've tried to explain in my previous comments. Now, again, our job is to help you with this as best we can, of course!

**Reply by axelboc** [reply]
As a first step to get you going, we can try to make a quick CodeSandbox for you with a small abstraction of an R3F `mesh` that accepts mouse callbacks that work in data space. You can then copy this component into daiquiri and see if it works as you expect. Once stabilised, we can bring it into H5Web. How does that sound?

**Comment by loichuder** [comment]
> it moves from right to left to optimize the location in the plot ; but i really don't want blinking of the tooltip content, the screen is not limited to a widget.

This is because we use visx's  `TooltipWithBounds`. Using visx's `Tooltip` deactivates the behaviour.
